code/datums/chatmessage.dm

```diff
@@ -313,6 +327,23 @@
 			return "#[num2hex(x, 2)][num2hex(m, 2)][num2hex(c, 2)]"
 		if(5)
 			return "#[num2hex(c, 2)][num2hex(m, 2)][num2hex(x, 2)]"
+/**
+ * Approximates the chatmesseges width based on cached widths of each char.
+ * If the character is not found in this cache we assume the worst and add the highest possible value.
```

Comment:
does this behave poorly with unicode and friends then?

Reply:
You shouldn't spam weird Unicode symbols anyway. If the symbol is bigger than biggest symbol (M) it may cut the message in some really edge cases.

Reply:
yeah but if people do are we gonna fuck them?

Reply:
😳

Reply:
Your code absolutely should be prepared for Unicode text, because not all Unicode text is emojis. People using accents, or even something like Chinese characters, are all things that the codebase should be prepared for.

Reply:
99% characters used in speak are not unicode characters. By assuming the largest pixel count (M which is about 10px) the only things that will cause actual issues are the weird super-long characters and spamming them is definitely malicious.
The worst that could happen is that our message will take one more line, which is small visual bug, that could happen already.

I guess if you could prepare some txt files with all the chars we want to track and read it and calculate our value based on those characters.

Reply:
> The worst that could happen is that our message will take one more line

That's fine then. Unicode characters *extending* the size is fine, since at least you can still see everything.

Reply:
I would be interested in seeing what this does to accent characters, and also kinda russian

Reply:
@Archanial a 100% safe solution is to do the approximation unless it scans a character that isnt in its list of font characters, in which case it gives up and uses MeasureText() on the message instead. as long as your list of characters covers 99.9% of speech it only falls back in the tiniest of edge cases

Reply:
Small Fonts *does* have the accented characters for the major european languages. It would be nice if it at least supported those.

![image](https://user-images.githubusercontent.com/1185434/170138195-caab8d51-c98f-4fc3-a25c-799d051b97b8.png)

Reply:
Also, it appears that this could be a global function since it doesn't use src at all as far as I can see, and then `balloon_alert_perform()` could also use it.

Or possibly also to make status displays seemlessly loop. Status displays only have the gaps after 3 repetitions because it was easier to do that than actually attempt to measure text when there's no users for them.

Reply:
On bee balloon alerts are really subtypes of chatmessages.
I could either repath it and implement the function there.
Or make the proc global and reuse it. 
I'll leave it up to maintainers if they want either.

Reply:
Do not make them a subtype of chat messages, make the proc global